File: src/main/java/jalation/graphalytics/DualIso.java

//::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::
/** @author  Matthew Saltz, John Miller, Ayushi Jain
 *  @version 1.1
 *  @date    Thu Jul 25 11:28:31 EDT 2013
 *  @see     LICENSE (MIT style license file).
 */

package jalation.graphalytics;

import java.util.Arrays;
import java.util.ArrayList;
import java.util.HashSet;
import java.util.List;
import java.util.Set;

import static java.lang.System.out;

//::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::
/** The `DualIso` class provides an implementation for Subgraph Isomorphism that
 *  uses Dual Graph Simulation for pruning.
 *
 *  @param g  the data graph G(V, E, l)
 *  @param q  the query graph Q(U, D, k)
 */
public class DualIso
{
    /** check progress after this many matches
     */
    private static final int CHECK = 1024;

    /** the data graph G(V, E, l)
     */
    private final Graph g;

    /** the query graph Q(U, D, k)
     */
    private final Graph q;

    /** the Dual Graph Simulation pattern matcher
     */
    private final DualSim2 duals;

    /** hold matches between vertices in q and g
     */
    private Set <Set <Integer> []> matches = null;

    /** indication of whether bijection have been created
     */
    private boolean noBijections = true;

    /** quit after too many matches (use setLimit () to adjust)
     */
    private int limit = 1000000;

    //::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::
    /** Construct a `DualIso` matcher for Subgraph Isomorphism.
     *
     *  @param _g  the data graph
     *  @param _q  the query graph
     */
    public DualIso (Graph _g, Graph _q)
    {
        g     = _g;
        q     = _q;
        duals = new DualSim2 (g, q);
    } // constructor

    //::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::
    /** Set the limit on the number of matches to find before quiting.
     *
     *  @param _limit  the limit on the number of matches to find
     */
    public void setLimit (int _limit)
    {
        limit = _limit;
    } // setLimit

    //::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::
    /** Apply the Dual Subgraph Isomorphism algorithm to find subgraphs of data
     *  graph 'g' that isomorphically match query graph 'q'.  These are
     *  represented by a set of single-valued bijective functions {'psi'} where
     *  each 'psi' function maps each query graph vertex 'u' to a data graph
     *  vertices 'v'.
     */
    public Set <Integer []> bijections ()
    {
        matches = new HashSet <> ();                       // initialize matches to empty
        Set <Integer> [] phi = duals.feasibleMates ();     // initial mappings from label match
        saltzDualIso (duals.saltzDualSim (phi), 0);        // recursively find all bijections
        Set <Integer []> psi = simplify (matches);         // pull bijections out matches
        noBijections = false;                              // results now available
        return psi;                                        // return the set of bijections
    } // bijections

    //::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::
    /** Apply the Dual Subgraph Isomorphism pattern matching algorithm to find
     *  the mappings from the query graph 'q' to the data graph 'g'.  These are
     *  represented by a multi-valued function 'phi' that maps each query graph
     *  vertex 'u' to a set of data graph vertices '{v}'.
     */
    public Set <Integer> [] mappings ()
    {
        Set <Integer []> psi = null;                   // mappings from Dual Simulation
        if (noBijections) psi = bijections ();         // if no results, create them
        return merge (psi);                            // merge bijections to create mappings
    } // mappings

    //::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::
    /** Return the count of the number of matches.
     */
    public int numMatches ()
    {
        return matches.size ();
    } // numMatches

    //::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::
    /** Refine the mappings 'phi' using the Dual Subgraph Isomorphism algorithm.
     *  Enumerate bijections by using an Ullmann-like recursion that uses Dual
     *  Graph Simulation for pruning.
     *
     *  @param phi    array of mappings from a query vertex u_q to { graph vertices v_g }
     *  @param depth  the depth of recursion
     */
    private void saltzDualIso (Set <Integer> [] phi, int depth)
    {
        if (depth == q.size ()) {
            if (phi.length > 0) {
                matches.add (phi);
                if (matches.size () % CHECK == 0) {
                    out.println ("dualIso: matches so far = " + matches.size ());
                } // if
            } // if

        } else if (phi.length > 0) {
            for (int i: phi [depth]) {
                if (! contains (phi, depth, i)) {
                    Set <Integer> [] phiCopy = copy (phi);                     // make a copy of phi
                    phiCopy [depth] = new HashSet <> (Arrays.asList (i));      // isolate vertex i
                    if (matches.size () >= limit) return;                      // quit if at LIMIT
                    saltzDualIso (duals.saltzDualSim (phiCopy), depth + 1);    // solve recursively for the next depth
                } // if
            } // for
        } // if
    } // saltzDualIso

    //::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::
    /** Copy the array of sets 'phi'.
     *
     *  @param phi  the array of sets to be copied
     */
    @SuppressWarnings("unchecked")
    private Set <Integer> [] copy (Set <Integer> [] phi)
    {
        Set <Integer> [] c = (Set <Integer> []) new Set <?> [phi.length];
        for (int i = 0; i < phi.length; i++) c[i] = new HashSet <> (phi [i]);
        return c;
    } // copy

    //::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::
    /** Is vertex j contained in any phi(i) for the previous depths?
     *
     *  @param phi    array of mappings from a query vertex u_q to { graph vertices v_g }
     *  @param depth  the current depth of recursion
     *  @param j      the vertex j to check
     */
    private boolean contains (Set <Integer> [] phi, int depth, int j)
    {
        for (int i = 0; i < depth; i++) {
            if (phi [i].contains (j)) return true;
        } // for
        return false;
    } // contains

    //::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::
    /** Create an array to hold matches for each vertex 'u' in the query graph
     *  'q' and initialize it to contain all empty sets.  Then for each bijection,
     *  add each element of the bijection to its corresponding match set.
     *
     *  @param psi  the set of bijections
     */
    @SuppressWarnings("unchecked")
    private Set <Integer> [] merge (Set <Integer []> psi)
    {
        Set <Integer> [] match = (Set <Integer> []) new Set <?> [q.size ()];
        for (Integer [] b: bijections ()) {
            for (int i = 0; i < b.length; i++) match [i].add (b[i]);
        } // for
        return match;
    } // merge

    //::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::
    /** Pull the bijections out of the complete match set.
     *
     *  @param matches  the complete match set embedding all bijections
     */
    private Set <Integer []> simplify (Set <Set <Integer> []> matches)
    {
        Set <Integer []> rset = new HashSet <> ();
        for (Set <Integer> [] m: matches) {
            Integer [] result = new Integer [q.size ()];
            for (int j = 0; j < m.length; j++) result [j] = m[j].iterator ().next ();
            rset.add (result);
        } // for
//      out.println ("rset = " + Arrays.deepToString (rset.toArray ()));
        return rset;
    } // simplify

    //::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::
    /** Test the `DualIso` class.
     *
     *  @param args  unused command-line arguments
     */
    @SuppressWarnings("unchecked")
    public static void main (String [] args)
    {
        int gSize    = 1000;       // size of the data graph
        int qSize    = 10;         // size of the query graph
        int nLabels  = 100;        // number of distinct labels
        int gAvDegree = 5;         // average vertex out degree for data graph
        int qAvDegree = 2;         // average vertex out degree for query graph

//      Graph g = GraphGenerator.genRandomGraph (gSize, nLabels, gAvDegree);
//      Graph q = GraphGenerator.genBFSQuery (qSize, qAvDegree, g);

        Set <Integer> [] adj_g = (Set <Integer> []) new Set <?> [4];
        adj_g [0] = new HashSet <> (Arrays.asList (1, 2));
        adj_g [1] = new HashSet <> (Arrays.asList (2, 3));
        adj_g [2] = new HashSet <> (Arrays.asList (3));
        adj_g [3] = new HashSet <> (Arrays.asList (1));
        Integer [] label_g = {2, 1, 2, 1};
        Graph g = new Graph (adj_g, label_g, false);
        g.print ();

//      Set <Integer> [] adj_q = (Set <Integer> []) new Set <?> [3];
//      adj_q [0] = new HashSet <> (Arrays.asList (1, 2));
//      adj_q [1] = new HashSet <> (Arrays.asList (2));
//      adj_q [2] = new HashSet <> ();
//      Integer [] label_q = {2, 1, 2};
//      Graph q = new Graph (adj_q, label_q, false);
//      q.print (); 

        Set <Integer>[] adj_q = (Set <Integer> []) new Set <?> [4];
        adj_q [0] = new HashSet <> (Arrays.asList (1, 2));
        adj_q [1] = new HashSet <> (Arrays.asList (2, 3));
        adj_q [2] = new HashSet <> (Arrays.asList (3));
        adj_q [3] = new HashSet <> (Arrays.asList (1));
        Integer [] label_q = {2, 1, 2, 1};
        Graph q = new Graph (adj_q, label_q, false);

        DualIso matcher = new DualIso (g, q);             // Dual Subgraph Isomorphism Pattern Matcher
        for (int i = 0; i < 3; i++) {
            out.println ("-------------------------------------------");
            g.print ();
            q.print ();
            Set <Integer []> psi = matcher.bijections ();                    // run the matcher
//          for (Integer [] p: psi) out.println (Arrays.toString (p));
            out.println ("Number of Matches: " + matcher.numMatches ());
        } // for
    } // main

} // DualIso class

File: scalation_database/src/main/scala/scalation/graph_db/pattern_matching/DualIso.scala

//::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::
/** @author  Matthew Saltz, John Miller, Ayushi Jain
 *  @version 1.4
 *  @date    Thu Jul 25 11:28:31 EDT 2013
 *  @see     LICENSE (MIT style license file).
 */

package scalation.graph_db
package pattern_matching

import scala.collection.mutable.Map
import scala.collection.mutable.{Set => SET}
import scala.reflect.ClassTag

import scalation.util.{banner, time}

//::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::
/** The `DualIso` class provides an implementation for Subgraph Isomorphism
 *  that uses Dual Graph Simulation for pruning.
 *  @param g      the data graph  G(V, E, l)
 *  @param q      the query graph Q(U, D, k)
 *  @param duals  the base dual simulation pattern matcher
 */
class DualIso [TLabel: ClassTag] (g: Graph [TLabel], q: Graph [TLabel], duals: GraphMatcher [TLabel])
      extends GraphMatcher (g, q)
{
    private var t0           = 0.0                         // start time for timer
    private var matches      = SET [Array [SET [Int]]] ()  // initialize matches to empty
    private var noBijections = true                        // no results yet
    private var limit        = 1000000                     // limit on number of matches

    //::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::
    /** Set an upper bound on the number matches to allow before quitting.
     *  @param _limit  the number of matches before quitting
     */
    def setLimit (_limit: Int) { limit = _limit }

    //::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::
    /** Apply the Dual Subgraph Isomorphism algorithm to find subgraphs of data
     *  graph 'g' that isomorphically match query graph 'q'.  These are represented
     *  by a set of single-valued bijections {'psi'} where each 'psi' unction
     *  maps each query graph vertex 'u' to a data graph vertices 'v'.
     */
    override def bijections (): SET [Array [Int]] =
    {
        matches = SET [Array [SET [Int]]] ()               // initialize matches to empty
        val phi = duals.feasibleMates ()                   // initial mappings from label match
        refine (duals.prune (phi), 0)                      // recursively find all bijections
        val psi = simplify (matches)                       // pull bijections out of matches
        noBijections = false                               // results now available
        psi                                                // return the set of bijections
    } // bijections

    //::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::
    /** Apply the Dual Subgraph Isomorphism pattern matching algorithm to find
     *  the mappings from the query graph 'q' to the data graph 'g'.  These are
     *  represented by a  multi-valued function 'phi' that maps each query graph
     *  vertex 'u' to a set of data graph vertices '{v}'.
     */
    override def mappings (): Array [SET [Int]] = 
    {
        var psi: SET [Array [Int]] = null              // mappings from Dual Simulation
        if (noBijections) psi = bijections ()          // if no results, create them
        merge (psi)                                    // merge bijections to create mappings
    } // mappings

    //::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::
    /** Return the count of the number of matches.
     */
    def numMatches (): Int = matches.size

    //::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::
    /** Refine the mappings 'phi' using the Dual Subgraph Isomorphism algorithm.
     *  Enumerate bijections by using an Ullmann-like recursion that uses Dual
     *  Graph Simulation for pruning.
     *  @param phi    array of mappings from a query vertex u_q to { graph vertices v_g }
     *  @param depth  the depth of recursion
     */
    private def refine (phi: Array [SET [Int]], depth: Int)
    {
        if (depth == q.size) {
            if (! phi.isEmpty) {
                matches += phi
                if (matches.size % CHECK == 0) println ("dualIso: matches so far = " + matches.size)
            } // if
        } else if (! phi.isEmpty) {
            for (i <- phi (depth) if (! contains (phi, depth, i))) {
                val phiCopy = phi.map (x => x)                           // make a copy of phi
                phiCopy (depth) = SET [Int] (i)                          // isolate vertex i
                if (matches.size >= limit) return                        // quit if at LIMIT
                refine (duals.prune (phiCopy), depth + 1)                // solve recursively for the next depth
            } // for
        } // if
    } // refine

    //::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::
    /** Determine whether vertex 'j' is contained in any 'phi(i)' for the previous depths.
     *  @param phi    array of mappings from a query vertex u_q to { graph vertices v_g }
     *  @param depth  the current depth of recursion
     *  @param j      the vertex j to check
     */
    private def contains (phi: Array [SET [Int]], depth: Int, j: Int): Boolean =
    {
        for (i <- 0 until depth if phi(i) contains j) return true
        false
    } // contains

    //::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::
    /** Create an array to hold matches for each vertex 'u' in the query graph
     *  'q' and initialize it to contain all empty sets.  Then for each bijection,
     *  add each element of the bijection to its corresponding match set.
     *  @param psi  the set of bijections
     */
    private def merge (psi: SET [Array [Int]]): Array [SET [Int]] =
    {
        val matches = Array.ofDim [SET [Int]] (q.size).map (_ => SET [Int] ())
        for (b <- bijections; i <- b.indices) matches(i) += b(i)
        matches
    } // merge

    //::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::
    /** Pull the bijections out of the complete match set.
     *  @param matches  the complete match set embedding all bijections
     */
    private def simplify (matches: SET [Array [SET [Int]]]): SET [Array [Int]] =
    {
        // FIX - may fail on .next method - provide a more robust implementation
        matches.map (m => m.map (set => set.iterator.next))
    } // simplify

   //::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::
    /** The 'prune' is not needed, pruning is delegated to incorporated graph
     *  simulation algorithm.
     *  @param phi  array of mappings from a query vertex u_q to { graph vertices v_g }
     */
    def prune (phi: Array [SET [Int]]): Array [SET [Int]] = throw new UnsupportedOperationException ()

    //::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::
    /** Count the number of mappings between query graph vertices 'u_i' and their sets
     *  of data graph vertices {v}, giving the number of distinct vertices and edges.
     *  @param psi  the set of bijection
     */
    def countMappings (psi: SET [Array [Int]]): Pair =
    {
        val distVertices = SET [Int] ()
        val distEdges    = SET [Pair] ()
        for (pi <- psi) distVertices ++= pi
        for (pi <- psi) {
            for (v <- pi; v_c <- g.ch(v) if distVertices contains v_c) distEdges += ((v, v_c))
        } // for
        (distVertices.size, distEdges.size)
    } // countMappings

    //::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::
    /** Test the Graph Iso Pattern Matcher.
     *  @param mName  the name of graph pattern matcher
     *  @param ans    the correct answer
     */
    def test (name: String, ans: SET [Array [Int]]): SET [Array [Int]] =
    {
        val psi = time { bijections () }                            // time the matcher
        banner (s"query ${q.name} (${q.size}) via $name on data ${g.name} (${g.size})")
//      for (pi <- psi) println (s"psi = ${pi.deep}")
        println (s"number of matches = $numMatches")
        println (s"(#distVertices, #distEdges) = ${countMappings (psi)}")

        if (ans != null) {
            // FIX - compare result with answer
        } // if 
        psi
    } // test

} // DualIso class

import scalation.graph_db.{ExampleGraphD => EX_GRAPH}

//::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::
/** The `DualIsoTest` object is used to test the `DualIso` class.
 *  > runMain scalation.graph_db.pattern_matching.DualIsoTest
 */
object DualIsoTest extends App
{
    val g = EX_GRAPH.g1
    val q = EX_GRAPH.q1

    println (s"g.checkEdges = ${g.checkEdges}")
    g.printG ()
    println (s"q.checkEdges = ${q.checkEdges}")
    q.printG ()

    val matcher = new DualIso (g, q, new DualSim2 (g, q))             // Dual Subgraph Isomorphism Pattern Matcher
    val psi = time { matcher.bijections () }                          // time the matcher
    println (s"number of matches = ${matcher.numMatches}")
    println (s"(vertices, edges) = ${matcher.countMappings (psi)}")
    for (p <- psi) println (s"psi = ${p.deep}")

} // DualIsoTest object


//::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::
/** The `DualIsoTest2` object is used to test the `DualIso` class.
 *  > runMain scalation.graph_db.pattern_matching.DualIsoTest2
 */
object DualIsoTest2 extends App
{
    val g = EX_GRAPH.g2
    val q = EX_GRAPH.q2

    println (s"g.checkEdges = ${g.checkEdges}")
    g.printG ()
    println (s"q.checkEdges = ${q.checkEdges}")
    q.printG ()

    val matcher = new DualIso (g, q, new DualSim2 (g, q))             // Dual Subgraph Isomorphism Pattern Matcher
    val psi = time { matcher.bijections () }                          // time the matcher
    println (s"number of matches = ${matcher.numMatches}")
    println (s"(vertices, edges) = ${matcher.countMappings (psi)}")
    for (p <- psi) println (s"psi = ${p.deep}")

} // DualIsoTest2 object


//::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::
/** The `DualIsoTest3` object is used to test the `DualIso` class.
 *  > runMain scalation.graph_db.pattern_matching.DualIsoTest3
 */
object DualIsoTest3 extends App
{
    val gGen = new GraphGen [Double] (0.0)

    val gSize     = 1000         // size of the data graph
    val qSize     =   10         // size of the query graph
    val nLabels   =  100         // number of distinct labels
    val gAvDegree =    5         // average vertex out degree for data graph
    val qAvDegree =    2         // average vertex out degree for query graph

    val g = gGen.genRandomGraph (gSize, nLabels, gAvDegree, false, "g")
    val q = gGen.genBFSQuery (qSize, qAvDegree, g, false, "q")

    println (s"q.checkEdges = ${q.checkEdges}")
    q.printG ()

    val matcher = new DualIso (g, q, new DualSim2 (g, q))             // Dual Subgraph Isomorphism Pattern Matcher
    val psi = time { matcher.bijections () }                          // time the matcher
    println (s"number of matches = ${matcher.numMatches}")
    println (s"(vertices, edges) = ${matcher.countMappings (psi)}")
    for (p <- psi) println (s"psi = ${p.deep}")

} // DualIsoTest3 object

